cnchi/mirrors.py

```diff
@@ -30,45 +30,324 @@
+        if row:
+            self.set_mirror_active(row.data, switch.get_active())
+            self.emit("switch-activated")
+
+    def on_drag_begin(self, widget, drag_context):
+        """ User starts a drag """
+        row = widget.get_ancestor(Gtk.ListBoxRow)
+        alloc = row.get_allocation()
+        surface = cairo.ImageSurface(cairo.FORMAT_ARGB32, alloc.width, alloc.height)
+        ctx = cairo.Context(surface)
+
+        row.get_style_context().add_class("drag-icon")
+        row.draw(ctx)
+        row.get_style_context().remove_class("drag-icon")
+
+        (x, y) = widget.translate_coordinates (row, 0, 0)
+
+        surface.set_device_offset(-x, -y)
+        Gtk.drag_set_icon_surface(drag_context, surface)
+
+    def on_drag_data_get(self, widget, drag_context, selection_data, info, time):
+        """ When drag data is requested by the destination """
+        row = widget.get_ancestor(Gtk.ListBoxRow)
+        listbox_str = str(self)
+        row_index = row.get_index()
+        data = "{0}|{1}".format(listbox_str, row_index)
+        selection_data.set_text(data, len(data))
+
+    def on_drag_data_received(self, widget, drag_context, x, y, selection_data, info, time):
+        """ When drag data is received by the destination """
+        data = selection_data.get_text()
+        try:
+            listbox_str = data.split('|')[0]
+            if listbox_str == str(self):
+                row_index = widget.get_index()
+                old_index = int(data.split('|')[1])
+                new_index = widget.get_index()
+                self.mirrors.insert(new_index, self.mirrors.pop(old_index))
+                self.fillme()
+        except (KeyError, ValueError) as err:
+            pass
+
+
+    def save_changes(self):
+        """ Save mirrors in mirrors list file """
+        # Save a backup if possible
+        src = self.mirrors_file_path
+        dst = src + ".cnchi-backup"
+
+        try:
+            shutil.copy2(src, dst)
+        except (FileNotFoundError, FileExistsError, OSError) as err:
+            logging.warning(err)
+
+        # ok, now save our changes
+        with open(src, 'w') as mfile:
+            line = "# Mirrorlist file modified by Cnchi\n\n"
+            mfile.write(line)
+            for (url, active) in self.mirrors:
+                line = "Server = {}\n".format(url)
+                if not active:
+                    line = "#" + line
+                mfile.write(line)
 
-        # Set up list box
-        self.listbox = self.ui.get_object("listbox")
-        self.listbox.set_selection_mode(Gtk.SelectionMode.NONE)
-        self.listbox.set_sort_func(self.listbox_sort_by_name, None)
 
-        # TODO: By default, select automatic mirror list ranking
+class Mirrors(GtkBaseBox):
+    def __init__(self, params, prev_page="features", next_page="installation_ask"):
+        super().__init__(self, params, "mirrors", prev_page, next_page)
+
+        data_dir = self.settings.get("data")
```

Comment:
__[flake8]__
* [F841] local variable 'data_dir' is assigned to but never used

<!-- Comment by [SideCI](https://sideci.com) -->